docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,106 @@
+
+# User experience
+
+## Expressing a requirement for chaining
+
+To express that the application requires chaining for self-requests, add the following to `spin.toml`:
+
+```toml
+[application.prerequisites]
+service_chaining = "required"
+```
+
+**BIKESHED:** Or something more explicit such as `direct_self_requests`.  Or `prerequisites` could be an array of capabilities that the host must provide (`prerequisites = ["service-chaining"]`).  Also, we _could_ call this something like `environment` or `host` instead, but I'm kind of inclined to leave it generic as to what prerequisites might go in here.
+
+Entries in `prerequisites` are considered must-understand.  For the section itself, that has to happen at a `serde` level since we can't retrofit it to existing builds.  But `serde` errors are not very lovely, and I'd prefer hosts in future to be able to say "hey I don't support your `gpu` requirement" rather than "unknown field serialisation error."
+
+## Requesting a component via chaining
+
+There are two ways we could express a HTTP request to a chained component:
+
+1. Use the existing self-request style.  The host _may_ chain self-requests, and _must_ chain them if the "require chaining" setting is on.  This is super easy for the application developer, at both the send and receive end.  However, it becomes more complex for the host, as the host now has to route the request - this drags in much more of the HTTP trigger plumbing.  (The prototype did this and it was egregiously hairy.)
+
+2. Address requests directly to the destination components by name. We still need a URL, so this would introduce a special host name such as `spin.internal`.  Then the URL `http://spin.internal/accounts` would invoke the `accounts` component.  This bypasses routing, which should simplify implementation.  It also opens the door to private endpoints, where an internal microservice is not exposed to the network (does not have a HTTP route) and so is accessible only from within the application.  However, chained components which parse URLs may need to take extra care to correctly handle both routed and chaining URLs.
+
+The preferred user experience is option 2: `http://spin.internal/<component-name>`
+
+**DETAIL NEEDED:** How will Spin request headers be set in this case?  One option is that Spin won't set them _at all_ and the calling component has to inject them if the receiver needs them.  Otherwise e.g. Consider a request to `http://spin.internal/accounts/314`.  `spin-path-info` (the 'wildcard match' bit of the path) is probably `/314` (bit after component name), but what about `spin-matched-route` or `spin-component-route`?  Are they just not set?  What of `spin-client-addr` - forwarded or skipped?
+
+**DETAIL NEEDED:** Should Spin automatically set an additional header indicating that this is a chained (unrouted) call?  If so, should it contain any additional data?  Or is it up to the caller to decide what, if anything, they want to disclose?
```

Comment:
I think that we should (at least) say that `host: *.spin.internal` appears on requests if and only if they have been routed internally.

Brainstorming, it might be nice to also define a header prefix that similarly can only traverse internal calls, i.e. would be filtered out of any inbound http request: something like `spin-internal-*`.